kcm should use the controller name constants

carlory opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

A follow-up of https://github.com/kubernetes/kubernetes/pull/120371

1. logging should use a canonical controller name when referencing a controller (Eg. Starting X, Shutting down X)
2. calling WaitForNamedCacheSync

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related-to https://github.com/kubernetes/kubernetes/pull/120371

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
All controllers in kcm used the controller name constants when
1. logging should use a canonical controller name when referencing a controller (Eg. Starting X, Shutting down X)
2. calling WaitForNamedCacheSync
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

carlory commented:
/cc @atiratree

carlory commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -310,7 +312,7 @@ func (c *Controller) sync(ctx context.Context, key string) error {
 					WithLastTransitionTime(metav1.Now()))
 			svcApply := networkingapiv1beta1apply.ServiceCIDR(cidr.Name).WithStatus(svcApplyStatus)
-			_, err = c.client.NetworkingV1beta1().ServiceCIDRs().ApplyStatus(ctx, svcApply, metav1.ApplyOptions{FieldManager: controllerName, Force: true})
+			_, err = c.client.NetworkingV1beta1().ServiceCIDRs().ApplyStatus(ctx, svcApply, metav1.ApplyOptions{FieldManager: c.name, Force: true})
```
- controllerName = "service-cidr-controller" https://github.com/kubernetes/kubernetes/pull/127550/files#diff-37ce781520331481491143bc84a058962599acdc003edf2e905cfa60fb52d8b9L57
- c.name = "service-cidr-controller" https://github.com/carlory/kubernetes/blob/ctrl-name/cmd/kube-controller-manager/names/controller_names.go#L85

FieldManager is not changed.

carlory commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -340,7 +342,7 @@ func (c *Controller) sync(ctx context.Context, key string) error {
 			WithLastTransitionTime(metav1.Now()))
 	svcApply := networkingapiv1beta1apply.ServiceCIDR(cidr.Name).WithStatus(svcApplyStatus)
-	if _, err := c.client.NetworkingV1beta1().ServiceCIDRs().ApplyStatus(ctx, svcApply, metav1.ApplyOptions{FieldManager: controllerName, Force: true}); err != nil {
+	if _, err := c.client.NetworkingV1beta1().ServiceCIDRs().ApplyStatus(ctx, svcApply, metav1.ApplyOptions{FieldManager: c.name, Force: true}); err != nil {
```
ditto

carlory commented on pkg/controller/validatingadmissionpolicystatus/controller.go:
```diff
@@ -156,6 +155,6 @@ func (c *Controller) reconcile(ctx context.Context, policy *v1.ValidatingAdmissi
 			WithTypeChecking(admissionregistrationv1apply.TypeChecking().
 				WithExpressionWarnings(warningsConfig...)))
-	_, err := c.policyClient.ApplyStatus(ctx, applyConfig, metav1.ApplyOptions{FieldManager: ControllerName, Force: true})
+	_, err := c.policyClient.ApplyStatus(ctx, applyConfig, metav1.ApplyOptions{FieldManager: c.name, Force: true})
```
ControllerName is `"validatingadmissionpolicy-status"` but the constant value is `"validatingadmissionpolicy-status-controller"`.

should I hardcode it with the previous value from ControllerName?

carlory replied:
/cc @jiahuif

carlory replied:
@atiratree How to deal with this problem? Can you give me some hints? Thanks a lot!

carlory commented:
retest

carlory commented:
/retest

carlory commented:
/cc @mborsz 

For endpointslice part

liggitt commented on pkg/controller/volume/pvprotection/pv_protection_controller_test.go:
```diff
@@ -210,7 +210,7 @@ func TestPVProtectionController(t *testing.T) {
 		// Create the controller
 		logger, _ := ktesting.NewTestContext(t)
-		ctrl := NewPVProtectionController(logger, pvInformer, client)
+		ctrl := NewPVProtectionController(logger, pvInformer, client, "persistentvolume-protection-controller")
```
shouldn't we pass in the constants to all these test invocations as well? hard-coding to a string gives the appearance that the name doesn't matter, when we actually want them to be consistent, right?

carlory replied:
@liggitt those constants are defined in `cmd/kube-controller-manager/names`, do we want to import this package into `pkg` and `tests` dir?

atiratree replied:
It is better not to depend on `/cmd`: https://github.com/kubernetes/kubernetes/blob/2fda3e4a650d8e729c509609aa865f0b6d843b4e/pkg/.import-restrictions#L8 

It would be nicer to use a constant, but I think it doesn't pose a big issue to use the hard-coded strings in the tests.

Hypothetically, does it make sense to modify the name to distinguish the test names? E.g. "persistentvolume-protection-controller-test"?

carlory replied:
> Hypothetically, does it make sense to modify the name to distinguish the test names? E.g. "persistentvolume-protection-controller-test"?

It's better to have the same value. It won't make developers confused when they look at the log file.

atiratree replied:
ok

carlory commented on pkg/controller/endpointslice/endpointslice_controller.go:
```diff
@@ -72,9 +72,9 @@ const (
 	// the component managing an EndpointSlice.
-	controllerName = "endpointslice-controller.k8s.io"
+	controllerQualifiedName = "endpointslice-controller.k8s.io"
 
```
Is this the normal form of controller names? Or is it a special case? @liggitt

liggitt replied:
this wasn't really rationalized with the controller name constants... it was only meant as a label value

most controllers don't set labels with their names

liggitt commented:
changing the event recorder sources is a user-visible change... does that matter? (are there clients checking events and looking for specific source strings?)

carlory commented:
> 是否有客户端检查事件并查找特定的源字符串？

I don't know. I can remove it from the release note if we don't want to document it.

liggitt commented:
> > are there clients checking events and looking for specific source strings?)
> 
> I don't know. I can remove it from the release note if we don't want to document it.

no, I'm saying we're making user-visible changes by propagating different constants into event sources ... I'm not sure whether that impact is ok

cici37 commented:
/triage accepted

atiratree commented on pkg/controller/bootstrap/bootstrapsigner.go:
```diff
@@ -159,15 +161,17 @@ func (e *Signer) Run(ctx context.Context) {
 
-	if !cache.WaitForNamedCacheSync("bootstrap_signer", ctx.Done(), e.configMapSynced, e.secretSynced) {
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting", "controller", e.name)
+	defer logger.Info("Shutting down controller", "controller", e.name)
```
This should be unified. Cc @pohly thoughts? 

```suggestion
	logger.Info("Starting", "controller", e.name)
	defer logger.Info("Shutting down", "controller", e.name)
```

or  

```suggestion
	logger.Info("Starting controller", "controller", e.name)
	defer logger.Info("Shutting down controller", "controller", e.name)
```

atiratree replied:
we also log similar thing here:

https://github.com/kubernetes/kubernetes/blob/3694a02bb9b5c1dd216485d2e25bd2c8f24916dd/cmd/kube-controller-manager/app/controllermanager.go#L766

maybe it makes sense to differentiate it somehow?

atiratree replied:
this is still unresolved, WDYT about the following proposal?

- let's do `Starting controller` and `Shutting down controller` in each controller
- change the `Starting controller` from the StartController function to `Initializing controller`

atiratree commented on pkg/controller/bootstrap/bootstrapsigner.go:
```diff
@@ -159,15 +161,17 @@ func (e *Signer) Run(ctx context.Context) {
+	logger.Info("Starting", "controller", e.name)
+	defer logger.Info("Shutting down controller", "controller", e.name)
+
+	if !cache.WaitForNamedCacheSync(e.name, ctx.Done(), e.configMapSynced, e.secretSynced) {
```
WaitForNamedCacheSync should be also changed to adhere to the `keysAndValues` logging format:

carlory replied:
@atiratree The whole package `cache` doesn't support the structured logging yet, so I prefer to do it in a follow-up PR. What do you think?

And it is tracked in the KEP as well: [contextual logging](https://github.com/kubernetes/enhancements/issues/3077)

atiratree replied:
ack, makes sense to do it later

atiratree commented on pkg/controller/cronjob/cronjob_controllerv2.go:
```diff
@@ -92,9 +93,10 @@ func NewControllerV2(ctx context.Context, jobInformer batchv1informers.JobInform
 		kubeClient:  kubeClient,
 		broadcaster: eventBroadcaster,
-		recorder:    eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "cronjob-controller"}),
+		recorder:    eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: controllerName}),
```
Maybe a better option would be to move the recorder changes to the followup PR and drive the discussion there.

atiratree commented on pkg/controller/endpointslice/endpointslice_controller.go:
```diff
@@ -72,9 +72,9 @@ const (
 	maxSyncBackOff = 1000 * time.Second
 
-	// controllerName is a unique value used with LabelManagedBy to indicated
+	// controllerQualifiedName is a unique value used with LabelManagedBy to indicated
```
```suggestion
	// controllerQualifiedName is a unique value used with LabelManagedBy to indicate
```

atiratree commented on pkg/controller/namespace/namespace_controller.go:
```diff
@@ -70,7 +72,8 @@ func NewNamespaceController(
 	resyncPeriod time.Duration,
-	finalizerToken v1.FinalizerName) *NamespaceController {
+	finalizerToken v1.FinalizerName,
+	controllerName string) *NamespaceController {
```
`controllerName` not passed further

atiratree commented on pkg/controller/validatingadmissionpolicystatus/controller.go:
```diff
@@ -66,12 +64,13 @@ func (c *Controller) Run(ctx context.Context, workers int) {
 		policyQueue: workqueue.NewTypedRateLimitingQueueWithConfig(
 			workqueue.DefaultTypedControllerRateLimiter[string](),
-			workqueue.TypedRateLimitingQueueConfig[string]{Name: ControllerName},
+			workqueue.TypedRateLimitingQueueConfig[string]{Name: controllerName},
```
We should not change names of the queues as it will affect the metrics they are reporting.

atiratree reviewed (commented):
@carlory thanks for picking this up!

carlory committed:
All controllers in kcm used the controller name constants when
1. logging should use a canonical controller name when referencing a controller (Eg. Starting X, Shutting down X)
2. calling WaitForNamedCacheSync

carlory commented:
> no, I'm saying we're making user-visible changes by propagating different constants into event sources ... I'm not sure whether that impact is ok

@liggitt I remove those changes from this PR.

carlory commented:
/test pull-kubernetes-kind-dra

atiratree commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -71,6 +71,8 @@ const (
 
 // Controller creates ResourceClaims for ResourceClaimTemplates in a pod spec.
 type Controller struct {
+	// name is the name of the controller
```
this comment is no longer needed

atiratree commented on pkg/controller/tainteviction/taint_eviction.go:
```diff
@@ -81,8 +81,7 @@ type GetPodsByNodeNameFunc func(nodeName string) ([]*v1.Pod, error)
 type Controller struct {
-	name string
-
+	controllerName        string
```
+1 for unification

atiratree commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -759,7 +759,7 @@ func StartController(ctx context.Context, controllerCtx ControllerContext, contr
 	}
 
-	logger.Info("Started controller", "controller", controllerName)
+	logger.Info("Initialized controller", "controller", controllerName)
```
And leave this line unchanged as well.

atiratree commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -728,12 +728,12 @@ func StartController(ctx context.Context, controllerCtx ControllerContext, contr
 	time.Sleep(wait.Jitter(controllerCtx.ComponentConfig.Generic.ControllerStartInterval.Duration, ControllerStartJitter))
 
-	logger.V(1).Info("Starting controller", "controller", controllerName)
+	logger.V(1).Info("Initializing controller", "controller", controllerName)
```
The though behind this is that we do initialization+starting here: `GetInitFunc`/`initFunc`. And the start itself will be logged by each controller.

carlory commented:
/test pull-kubernetes-e2e-kind-ipv6

atiratree commented:
/lgtm
/approve

cc @liggitt @thockin for the final review/approve

pohly commented on cmd/kube-controller-manager/app/apps.go:
```diff
@@ -43,6 +43,7 @@ func newDaemonSetControllerDescriptor() *ControllerDescriptor {
 func startDaemonSetController(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller.Interface, bool, error) {
 	dsc, err := daemon.NewDaemonSetsController(
 		ctx,
+		controllerName,
```
This is unnecessary because each logger gets instantiated with a context where the controller name is already included via `WithName`.

pohly replied:
/lgtm cancel

carlory replied:
```go
// nameKey is used to log the `WithName` values as an additional attribute.
nameKey = "logger"
```

The key is different when printing the controller name.

```go
	logger := klog.FromContext(ctx)
	logger.Info("Starting daemon sets controller")
	defer logger.Info("Shutting down daemon sets controller")
```

After I look at the code, there're some comments.

```
// In StartController, an explicit "controller" key is used instead, for two reasons:
// - while contextual logging is alpha, klog.LoggerWithName is still a no-op,
//   so we cannot rely on it yet to add the name
// - it allows distinguishing between log entries emitted by the controller
//   and those emitted for it - this is a bit debatable and could be revised.
```

@pohly  What's current state of the contextual logging feature? And What's the desired key name for controller name, `logger` or `controller`?

Where should other todos get the controller name?

```
//  2. used anywhere inside the controller itself:
//     2.1. [TODO] logging should use a canonical controller name when referencing a controller (Eg. Starting X, Shutting down X)
//     2.2. [TODO] emitted events should have an EventSource.Component set to the controller name (usually when initializing an EventRecorder)
//     2.3. [TODO] registering ControllerManagerMetrics with ControllerStarted and ControllerStopped
//     2.4. [TODO] calling WaitForNamedCacheSync
```

- 2.2 may need to move initializing an EventRecorder into `Run` func. it's tracked in https://github.com/kubernetes/kubernetes/issues/128282
- 2.3 may need to define a new func for each controller to run, i.e. `(c *Controller) RunWithNamedMetircs(ctx, controllerName, metrics)`
- 2.4 maybe able to get logger from ctx. It requires a new function. i.e. `WaitForCacheSyncWithContext`?

I'm unsure if my idea is correct.

carlory replied:
Great work! This PR can be closed. Thanks for your review!

For the first to-do, I think the remaining thing is to remove the `TODO`  and `(Eg. Starting X, Shutting down X)` words.

carlory replied:
replaced by https://github.com/kubernetes/kubernetes/pull/128537

/close

atiratree replied:
@pohly @carlory 

I think it is still useful to have the name passed separately:
- There is no way how to get the controllerName from the context. And it is also more readable to pass it directly.
- The logger name can be prefixed multiple times, and it is good to have a stable identifier for a controller:

The stable identifier is mainly needed for:
- clienBuilder (service account)
- eventSource.component
- queue names
- metrics / controllerManagerMetrics
- as a fieldManager


We do not have to duplicate the name in Starting/Shutting down, but IMO it is still useful to have when debugging the controllers. Because there may be a longer name path or an incorrect/incomplete path - depending under what context the controller is run.

For example in https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/certificates.go we run multiple controllers under the same context.

I think it would be good to reopen this PR and at least unify the messages and change the Apply calls.

atiratree replied:
> Each controller hard-coding its own name and potentially disagreeing with its caller on what that name is?

Yeah, most of the controllers have inconsistent hardcoded names for everything and it is difficult for us to change most of the stuff. But the new controllers should name everything consistently.

> Would it make sense to pass the controller name in the `ControllerContext`?

There is only one instance of ControllerContext per KCM and is used to pass `SharedInformerFactory`, etc. So it is not well suited for that use case.

> Whatever way the name is passed, let's clearly document that logging doesn't need to replicate the name because it is already in the contextual logger.

I am okay with that. The users will have to adapt to this change though.

pohly replied:
> There is only one instance of ControllerContext per KCM

It could get copied and modified specifically for each controller. But it's not important.

carlory commented:
/reopen

because the https://github.com/kubernetes/kubernetes/pull/127550#discussion_r1827729473 is still open.

carlory commented:
/hold
